test/sagemaker_tests/pytorch/training/resources/apex/ddp_race_condition_test.py

```diff
@@ -0,0 +1,78 @@
+# adapted from https://github.com/NVIDIA/apex/blob/master/tests/distributed/DDP/ddp_race_condition_test.py
+
+import torch
+import torch.distributed as dist
+from torch.nn import Parameter
+from torch.nn import Module
+from apex.parallel import DistributedDataParallel as DDP
+import argparse
+import os
+
+
+parser = argparse.ArgumentParser(description="allreduce hook example")
+parser.add_argument("--local_rank", default=0, type=int)
+args = parser.parse_args()
+
+dist.init_process_group(backend="nccl", init_method="env://")
+args.world_size = dist.get_world_size()
+args.local_rank = dist.get_rank()
+args.gpu = args.local_rank % torch.cuda.device_count()
+torch.cuda.set_device(args.gpu)
+print("assigned dist rank data", args.local_rank, args.world_size)
+
+torch.set_printoptions(precision=10)
+torch.manual_seed(args.local_rank)
+
+
+class Model(Module):
+    def __init__(self):
+        super(Model, self).__init__()
+        self.a = Parameter(torch.cuda.FloatTensor(4096 * 4096).fill_(1.0))
+        self.b = Parameter(torch.cuda.FloatTensor(4096 * 4096).fill_(2.0))
+
+    def forward(self, input):
+        return (input * self.a) * self.b
+
+
+model = Model()
+# model = DDP(model, message_size=1, gradient_predivide_factor=8.0)
+# model = DDP(model, delay_allreduce=True)
+# model = DDP(model, message_size=1, allreduce_trigger_params=[model.b])
+model = DDP(model, message_size=1, allreduce_trigger_params=[model.b], num_allreduce_streams=3)
+
+x = torch.cuda.FloatTensor(4096 * 4096)
+
+passed = True
+torch.cuda.cudart().cudaProfilerStart()
+for i in range(10):
+    x.fill_(i + args.local_rank)  # fill x with new values every iteration for sanity
+    model.zero_grad()
+    out = model(x)
+    loss = out.sum()
+    # torch.cuda.nvtx.range_push("backward")
+    loss.backward()
+    # torch.cuda.nvtx.range_pop()
+
+    # torch.cuda.nvtx.range_push("synchronize() + info")
+    # torch.cuda.synchronize()
+    print("i = {}".format(i))
+
+    def info(name, param, val):
+        expected = val * 4096 * 4096 * (2.0 * i + 1) / 2.0
+        actual = param.grad.data.sum().item()
+        print(
+            name
+            + ": grad.data_ptr() = {}, expected sum {}, got {}".format(
+                param.grad.data_ptr(), expected, actual
+            )
+        )
+        return expected == actual
+
+    if not info("model.a", model.module.a, 2.0):
```

Comment:
Nit: These two conditionals can be condensed into one, unless there is a specific case for each of these, in which case that should be logged. Do we want to also hard stop the test after the failed pass?

Reply:
I think it would make debugging easier to have these two function calls in separate lines of code.

By running both `module.a` and `module.b` regardless of whether the other failed will give us results from both tests, and a difference in success/failure between them both will give us more information than just one test.

Reply:
We might want to log that a certain module failed and hard fail at that point, if this were to fail after the fact it wouldn't provide any context into which model failed, and would not save any additional time.